src/lib/promise/js/promise.js

```diff
@@ -0,0 +1,147 @@
+// Provides: deferred_run
+function deferred_run(func) {
+  var deferred = {
+    promise: Promise.resolve()
+      .then(func) // the ocaml types don't know this, but func can actually be async or sync
+      .then(function (value) {
+        deferred.value = value;
+        deferred.isDetermined = true;
+        return value;
+      })
+      .catch(function (err) {
+        deferred.error = err;
+        deferred.isDetermined = true;
+        throw err;
+      }),
+    isDetermined: false,
+  };
+  return deferred;
+}
+
+// Provides: deferred_map
+function deferred_map(deferred, func) {
+  var newDeferred = {
+    promise: deferred.promise
+      .then(func) // the ocaml types don't know this, but func can actually be async or sync
+      .then(function (value) {
+        newDeferred.value = value;
+        newDeferred.isDetermined = true;
+        return value;
+      })
+      .catch(function (err) {
+        newDeferred.error = err;
+        newDeferred.isDetermined = true;
+        throw err;
+      }),
+    isDetermined: false,
+  };
+  return newDeferred;
+}
+
+// Provides: deferred_bind
+function deferred_bind(deferred, func) {
+  var newDeferred = {
+    promise: deferred.promise
+      .then(func)
+      .then(function (anotherDeferred) {
+        return anotherDeferred.promise;
+      })
+      .then(function (value) {
+        newDeferred.value = value;
+        newDeferred.isDetermined = true;
+        return value;
+      })
+      .catch(function (err) {
+        newDeferred.error = err;
+        newDeferred.isDetermined = true;
+        throw err;
+      }),
+    isDetermined: false,
+  };
+  return newDeferred;
+}
+
+// Provides: deferred_upon
+function deferred_upon(deferred, func) {
+  deferred.promise.then(function () {
+    func(deferred.value);
+  });
+}
+
+// Provides: deferred_is_determined
+function deferred_is_determined(deferred) {
+  return deferred.isDetermined;
+}
+
+// Provides: deferred_peek
+function deferred_peek(deferred) {
+  if (deferred.isDetermined) {
+    return [1, deferred.value];
```

Comment:
This doesn't handle the `error` case, which will lead to the (potentially junk) value `[1, undefined]`.

Reply:
The actual value in the case of an error ought to be `[0]` (ie. `None`, as below), for reference.

Reply:
I tested this now: actually `None` becomes `0` while `Some value` becomes `[1, value]`. Pretty weird...

Reply:
Ah yes, constructors with and without arguments are indexed separately in the OCaml compiler.

Reply:
Fixed